src/disc/disc.py

```diff
@@ -0,0 +1,249 @@
+    rid: str,
+    id: str = guild_id
+) -> str:
+    roles = await get_all_roles(id)
+    return roles[rid]
+
+async def get_role_by_id(
+    rid: str,
+    id: str = guild_id
+) -> list:
+    tok = os.environ.get('TOKEN')
+    url = f'https://discord.com/api/v10/guilds/{id}/roles'
+    result = await discord_request(url, tok)
+
+    json_s = result.json()
+    return list(filter(lambda x: x['id'] == guild_id, json_s))
+
+async def get_user_roles(
+    uid: str,
+    id: str = guild_id
+) -> list[str]:
+    tok = os.environ.get('TOKEN')
+    url = f'https://discord.com/api/v10/guilds/{id}/members/{uid}'
+    result = await discord_request(url, tok)
+
+    json_s = result.json()
+    return json_s['roles']
+
+
+async def get_all_roles(
+    id: str = guild_id
+) ->  dict[str, str, list[str]]:
+    tok = os.environ.get('TOKEN')
+    url = f'https://discord.com/api/v10/guilds/{id}/roles'
+    result = await discord_request(url, tok)
+
+    json_s = result.json()
+    roles = list(map(lambda x: ([x['id'], [x['name'], x['permissions']]]), json_s))
+    return dict(roles)
+
+async def get_guild_members_with_role(
+    rid: str,
+    id: str = guild_id
+) -> list[str]:
+    tok = os.environ.get('TOKEN')
+    # base case
+    url = f'https://discord.com/api/v10/guilds/{id}/members?limit=1'
+    result = await discord_request(url, tok)
+
+    json_s = result.json()
+    res = list(map(lambda x: [x['user']['username'], x['user']['id'], x['roles']], json_s))
+    match = list(filter(lambda x: rid in x[2], res))
+    matched = list(map(lambda x: x[0], match))
+    last_uid = res[-1][1]
+
+    # loop
+    while True:
+        url = f'https://discord.com/api/v10/guilds/{id}/members?limit=1&after={last_uid}'
+        result = await discord_request(url, tok)
+
+        json_s = result.json()
+        res = list(map(lambda x: [x['user']['username'], x['user']['id'], x['roles']], json_s))
+        match = list(filter(lambda x: rid in x[2], res))
+        match = list(map(lambda x: x[0], match))
+        matched = [*matched, *match]
+
+        if res == []:
+            break
+
+        last_uid = res[-1][1]
+    return matched
+    
+async def get_guild_members(
+    id: str = guild_id
+) -> list[list]:
+    tok = os.environ.get('TOKEN')
+    # base case
+    url = f'https://discord.com/api/v10/guilds/{id}/members?limit=1000'
+    result = await discord_request(url, tok)
+
+    json_s = result.json()
+    users = list(map(lambda x: [x['user']['username'], x['user']['id'], x['roles']], json_s))
+    last_uid = users[-1][1]
+
+    # loop
+    while True:
+        url = f'https://discord.com/api/v10/guilds/{id}/members?limit=1000&after={last_uid}'
+        result = await discord_request(url, tok)
+
+        json_s = result.json()
+        res = list(map(lambda x: [x['user']['username'], x['user']['id'], x['roles']], json_s))
+        users = [*users, *res]
+
+        if res == []:
+            break
+
+        last_uid = users[-1][1]
+    return users
```

Comment:
Since I had a lot of suggestions for this function, I've just re-written it using them. I think it would have been better for us to go through it together in-person, but alas. This function minimizes repeated code (construction of users tuple) and has an explicit exit condition for the loop.

NOTE: this assumes the default behaviour of `https://discord.com/api/v10/guilds/{id}/members` when there are no more members left to get, returns an empty list. I'm not 100% sure this is what occurs (api doc doesn't say), but it seems that way from the past version.

```suggestion
async def get_guild_members(
    id: str = guild_id
) -> list[list]:
    token = os.environ.get('TOKEN')
    users = []
    last_uid = None

    # base case
    url = f'https://discord.com/api/v10/guilds/{id}/members?limit=1000'
    result_json = await discord_request(url, token).json()

    # when there are no more members left to get, result_json will be an empty list.
    while len(result_json) != 0:
        users += [
            (member['user']['username'], member['user']['id'], member['roles'])
            for member in result_json
        ]
        last_uid = users[-1][1]

        url = f'https://discord.com/api/v10/guilds/{id}/members?limit=1000&after={last_uid}'
        result_json= await discord_request(url, token).json()

    return users
```